Katello/katello
# Fixes #35948 - Support for incremental syncable exports

#### What are the changes introduced in this pull request?
This commit adds the ability to do incremental syncable exports. 
We allowed only complete before for syncable and now you can do incremental. 

```
$ hammer content-export incremental repository --id=1 --format=syncable
[....................................................................................................................................................................] [100%]
Generated /var/lib/pulp/exports/Default_Organization/Export-SYNCABLE-interesting-1/12.0/2023-01-13T22-59-47-00-00/metadata.json
```

The files are exported in a syncable format but only contain rpms + metadata that changed since the last time you exported. 
cd into `/var/lib/pulp/exports/Default_Organization/Export-SYNCABLE-interesting-1/12.0/2023-01-13T22-59-47-00-00/` and you should see a directory structure that looks like the following. (In my repo I added a new duck rpm and it had gotten regenerated correctly and so you'd only notice 1 rpm)
```
$ cd /var/lib/pulp/exports/Default_Organization/Export-SYNCABLE-interesting-1/12.0/2023-01-13T22-59-47-00-00/
$ find custom
custom
custom/media
custom/media/interesting
custom/media/interesting/Packages
custom/media/interesting/Packages/d
custom/media/interesting/Packages/d/duck-0.8-1.noarch.rpm
custom/media/interesting/.treeinfo
custom/media/interesting/repodata
custom/media/interesting/repodata/68b9c4b0c0163b650160d76f691fd34cb1a2f18664b0136b965583ab3ad58640-primary.xml.gz
custom/media/interesting/repodata/e4ec28d7d0482b6db237b9a0a756fe704e28ca9bc73405b290e75d9723613848-filelists.xml.gz
custom/media/interesting/repodata/fcdaaa6cd37b259a599f01fc38f740ac50edb85de25227acb286b55797ec25bd-other.xml.gz
custom/media/interesting/repodata/3ae1b862f0d906e0f0f268bbccb16346c631d3dd6ed656fbf5dcba29c6f1f8bb-updateinfo.xml.gz
custom/media/interesting/repodata/540ff6c3a42ee01ab43f1d87b7b09d8d8787c5a505baa2a8764db31adc76c3d7-modules.yaml
custom/media/interesting/repodata/4eb81ce725d3a48773d1497d93acf9aed1cf2aa11da24ba228095b6edc372cba-comps.xml
custom/media/interesting/repodata/repomd.xml
```

You can then use this to sync or import into a downstream server via hammer content-import

##### Other Changes  #####
- Adds a setting for pulpcore_export_format which you can specify syncable/importable by default (so don't need to use --format)
- DRY'es up the content-export and content-incremental-export controllers (they had mostly duplicated descriptions and content) and puts the common code in exports-controller (which gets included in both controllers.)

#### Considerations taken when implementing this change?
So this is useful for a customer who wants to export repos in a syncable format.
Pulpcore PR => https://github.com/pulp/pulpcore/pull/3511 now adds the ability to do incremental syncable exports
Katello is taking advantage of this feature to implement incremental `syncable` exports.

All three content-export options are available to be extracted in syncable format.
- `hammer content-export incremental version --format=syncable ....`
- `hammer content-export incremental library --format=syncable ....`
- `hammer content-export incremental repository --format=syncable ....`

The export rpm files are generated as hardlinks to the actual file on disk. This there by avoids the 2X space we may need for regular importable exports.  

#### What are the testing steps for this pull request?
- On your dev env you may need to run `sudo setfacl --recursive --modify u:vagrant:rwX,d:u:vagrant:rwX /var/lib/pulp/exports` so that listing files can be generate when hammer is running as a vagrant user.
- On your katello instance `Administer -> Settings -> Content`. 
- set both `Default Custom Repository download policy` and `Default Red Hat Repository download policy` to immediate
- Create an org
- Add/Enable + sync few repos. Custom repos will be easier to check but RH should work just the same. 
- Create a content view with these repos an publish.
- Do a syncable complete export. One of
  - `hammer content-export complete version --format=syncable ....`
  - `hammer content-export complete library --format=syncable ....`
  - `hammer content-export complete repository --format=syncable ....`

For example:
```
$ hammer content-export complete repository --organization=org --product=prod --name=repo
[....................................................................................................................................................................] [100%]
Generated /var/lib/pulp/exports/org/Export-SYNCABLE-repo-19/1.0/2023-01-14T20-26-50-00-00/metadata.json
$ find /var/lib/pulp/exports/org/Export-SYNCABLE-repo-19/1.0/2023-01-14T20-26-50-00-00/ |grep rpm|wc
     27      27    3643
# 27 rpms in this list
```

- Upload an rpm to the custom repo so that a new version is created
 - Publish the content-view again
 Try any of the 3 hammer commands listed here
- `hammer content-export incremental version --format=syncable ....`
- `hammer content-export incremental library --format=syncable ....`
- `hammer content-export incremental repository --format=syncable ....`
```
$ hammer content-export incremental repository --organization=org --product=prod --name=repo --format=syncable
[....................................................................................................................................................................] [100%]
Generated /var/lib/pulp/exports/org/Export-SYNCABLE-repo-19/2.0/2023-01-15T04-23-20-00-00/metadata.json

$ find /var/lib/pulp/exports/org/Export-SYNCABLE-repo-19/2.0/2023-01-15T04-23-20-00-00/|grep rpm|wc
      1       1     130
```

- Make sure you only see the added rpm and new repo metadata
